libs/shared/flagd-core/src/lib/flagd-core.ts

```diff
@@ -91,35 +92,47 @@ export class FlagdCore {
       }
     }
 
-    const defaultVariant = flag.defaultVariant;
-
-    let resolvedVariant;
+    let variant;
     let reason;
 
-    if (flag.targetingString == undefined) {
-      resolvedVariant = flag.variants.get(defaultVariant);
+    if (!flag.targeting) {
+      variant = flag.defaultVariant;
       reason = StandardResolutionReasons.STATIC;
     } else {
-      // todo - targeting evaluation
-      // till targeting is handled, return the static result
-      resolvedVariant = flag.variants.get(defaultVariant);
-      reason = StandardResolutionReasons.STATIC;
+      let targetingResolution;
+      try {
+        targetingResolution = this._targeting.applyTargeting(flagKey, flag.targeting, evalCtx);
+      } catch (e) {
+        console.error(`Error evaluating targeting rule for flag ${flagKey}, falling back to default`, e);
+        targetingResolution = null;
+      }
+
+      if (!targetingResolution) {
+        variant = flag.defaultVariant;
+        reason = StandardResolutionReasons.DEFAULT;
+      } else {
+        variant = targetingResolution;
+        reason = StandardResolutionReasons.TARGETING_MATCH;
+      }
+    }
+
+    if (typeof variant !== 'string') {
+      throw new TypeMismatchError('Variant must be a string, but found ' + typeof variant)
     }
 
-    // todo improve this error condition when targeting evaluation is ready
-    if (resolvedVariant === undefined) {
-      throw new TypeMismatchError(`variant ${defaultVariant} not found in flag with key ${flagKey}`);
+    const resolvedVariant = flag.variants.get(variant)
+    if (!resolvedVariant) {
+      throw new TypeMismatchError(`Variant ${variant} not found in flag with key ${flagKey}`);
     }
 
     if (!guard(resolvedVariant)) {
-      throw new TypeMismatchError('evaluated type does not match the flag type');
+      throw new TypeMismatchError('Evaluated type does not match the flag type');
```

Comment:
It may be challenging to troubleshoot the problem with this message.

Could you please extend it to include information such as:
 - flag key
 - expected type
 - actual type

Reply:
Added suggested details to the error message